src/core/thread/mesh_forwarder.cpp

```diff
@@ -152,6 +153,102 @@ void MeshForwarder::Stop(void)
+        {
+            // For traffic destined to multicast address larger than realm local, generally it uses IP-in-IP
+            // encapsulation (RFC2473), with outer destination as ALL_MPL_FORWARDERS. So here if the destination
+            // is multicast address larger than realm local, it should be for indirection transmission for the
+            // device's sleepy child, thus there should be no direct transmission.
+            if (!ip6Header.GetDestination().IsMulticastLargerThanRealmLocal())
+            {
+                // schedule direct transmission
+                aMessage.SetDirectTransmission();
+            }
+
+            if (aMessage.GetSubType() != Message::kSubTypeMplRetransmission)
+            {
+                if (ip6Header.GetDestination() == mle.GetLinkLocalAllThreadNodesAddress() ||
+                    ip6Header.GetDestination() == mle.GetRealmLocalAllThreadNodesAddress())
+                {
+                    // destined for all rx-off children
+                    for (Child &child : Get<ChildTable>().Iterate(Child::kInStateValidOrRestoring))
+                    {
+                        if (!child.IsRxOnWhenIdle())
+                        {
+                            mIndirectSender.AddMessageForSleepyChild(aMessage, child);
+                        }
+                    }
+                }
+                else
+                {
+                    // destined for some sleepy children which subscribed the multicast address.
+                    for (Child &child : Get<ChildTable>().Iterate(Child::kInStateValidOrRestoring))
+                    {
+                        if (!child.IsRxOnWhenIdle() && child.HasIp6Address(ip6Header.GetDestination()))
+                        {
+                            mIndirectSender.AddMessageForSleepyChild(aMessage, child);
+                        }
+                    }
+                }
+            }
+        }
+        else if ((rxOffChild = Get<ChildTable>().FindChild(ip6Header.GetDestination())) != nullptr &&
+                 !rxOffChild->IsRxOnWhenIdle() && !aMessage.GetDirectTransmission())
+        {
+            // destined for an rx-off child
+            mIndirectSender.AddMessageForSleepyChild(aMessage, *rxOffChild);
+        }
+        else
+        {
+            // schedule direct transmission
+            aMessage.SetDirectTransmission();
+        }
+
+        break;
+    }
+#if OPENTHREAD_FTD
+    case Message::kTypeSupervision:
+    {
+        Child *child = Get<Utils::ChildSupervisor>().GetDestination(aMessage);
+        OT_ASSERT((child != nullptr) && !child->IsRxOnWhenIdle());
+        mIndirectSender.AddMessageForSleepyChild(aMessage, *child);
+        break;
+    }
+#endif
+    default:
+        aMessage.SetDirectTransmission();
+        break;
+    }
+#else
+    aMessage.SetDirectTransmission();
+#endif // OPENTHREAD_FTD || OPENTHREAD_MTD_S2S
+
+    mScheduleTransmissionTask.Post();
+
+    return OT_ERROR_NONE;
```

Comment:
I think for future, we can consider making this method `void`. We did not return error (other than `NONE` from FTD version either).